Add simple block appearance API

amadornes opened the pull request:
The proposed API allows blocks to report looking like a different block on a given face. This allows blocks with connected textures to seamlessly connect to other blocks that mimic them, such as the covers and facades often used to hide cables, without adding any extra dependencies and special-casing for each mod that adds such blocks.

An API of this style has been requested several times over the years, and has been previously denied due to the impracticality of their implementation and performance implications. However, Forge and Java have both matured plenty since then, making this a trivial addition with a negligible performance impact in the few cases it's used.

*This PR was made on request from Soaryn and the AE2 team.*

amadornes committed:
Add simple block appearance API

amadornes committed:
Enhance javadoc by specifying a method to safely pass in data from a BE

sciwhiz12 commented on src/main/java/net/minecraftforge/common/extensions/IForgeBlockState.java:
```diff
@@ -749,4 +749,17 @@ default boolean canBeHydrated(BlockGetter getter, BlockPos pos, FluidState fluid
+     * @param side  The side of the block that is being queried
+     * @return The appearance of this block from the given side
+     */
+    default BlockState getAppearance(BlockGetter level, BlockPos pos, Direction side)
```
This should have a `@see` tag directing to the `IForgeBlock`-equivalent method.

eerussianguy commented:
It may be worth noting in the javadoc that the method itself doesn't actually *do* anything, this is just a place to specify a common implementation for others to consume.

amadornes committed:
Update javadoc

amadornes committed:
That never happened

amadornes committed:
Fix javadoc again

Don't mind me, I'm just having a stroke over here

Technici4n commented on src/main/java/net/minecraftforge/common/extensions/IForgeBlock.java:
```diff
@@ -920,4 +923,27 @@ default MaterialColor getMapColor(BlockState state, BlockGetter level, BlockPos
+     * @param side  The side of the block that is being queried
+     * @return The appearance of this block from the given side
+     */
+    default BlockState getAppearance(BlockState state, BlockGetter level, BlockPos pos, Direction side)
```
Why not `BlockAndTintGetter` ?

amadornes committed:
Add query state and position by popular request and update javadoc

Also ensure that RenderChunkRegion returns a ModelDataManager, which for some reason it didn't. Without that, this whole thing wouldn't work.

Technici4n reviewed (approved):
Looks great, hope this can be merged soon. :shipit:
